### PR TITLE
PHPDOC fix: $object can be callable

### PR DESCRIPTION
**Pull request by ksimka**

### CODE REVIEW

File: src/Dispatcher.php [hunk 1]
```diff
@@ -103,7 +103,7 @@ public function __invoke($params = [], $object_name = null)
      * with the dispatchable method, a closure, or an invokable object,
      * recursively to that result with the same params.
      * 
-     * @param mixed $object Dispatch to this object.
+     * @param object|callable $object Dispatch to this object.
```
Comment (harikt) [inline src/Dispatcher.php]: This sounds mixed itself to me.

Comment (ksimka) [inline src/Dispatcher.php]: Oh, yes, you're right.

**Comment by pmjones** [comment]
@ksimka at first i thought this was a good catch, but after thinking about it, `callable` may not actually be correct. The logic in the dispatcher requires an object proper (or a Closure specifically). If you pass a function name, or an `array($object, $method)` then I don't think the Dispatcher will work properly.

Do you have further thoughts on that?

**Comment by ksimka** [comment]
@pmjones I didn't think about callable array, just wanted some proper highlight in IDE :)

But we can make dispatcher work with array with slight changes. Smth like this

``` php
    protected function dispatch($object, $params = [])
    {
        $method = $this->getMethodByParams($params);
        if (is_callable([$object, $method]) && method_exists($object, $method)) {
            // the object has the specified method
            $result = $this->invokeMethod($object, $method, $params);
        } elseif ($object instanceof Closure) {
            // the object is a closure proper
            $result = $this->invokeClosure($object, $params);
        } elseif (is_callable($object)) {
            if (is_object($object)) {
                $method = '__invoke';
            } elseif (is_array($object)) { // actually, is_array is not necessary here
                list($object, $method) = $object;
            }
            // the object is invokable
            $result = $this->invokeMethod($object, $method, $params);
        } else {
            // cannot dispatch any further; end recursion and return as-is
            return $object;
        }

        // recursively dispatch the result.
        return $this->dispatch($result, $params);
    }
```

If you want I can edit current patch.

**Comment by pmjones** [comment]
Appealing! However, it does mean that the `$method` ends up being completely ignored, which is not something I expect.

Is the array-format of a callable something we actually need with the Dispatcher? Honestly I have a hard time seeing that.

**Comment by ksimka** [comment]
> Appealing! However, it does mean that the $method ends up being completely ignored, which is not something I expect.

Yes, but we also override `__invoke` by `method` from params (if object has both), which is not obviuos too. You are the author, so you decide what is right :) If it is problem, you can add some kind of option to dispatcher that will configure priority of method's source.

And by the way, I can't imagine situation when programmer provides callable array as something that must be called and in the same time he wants `method` from params to be called. If so, why not use only object (first item of callable array) and put `method` to params?

> Is the array-format of a callable something we actually need with the Dispatcher? Honestly I have a hard time seeing that.

Yes, we already can do the same thing using 

``` php
$dispatcher(['service' => $service, 'method' => $method]);
```

instead of 

``` php
$dispatcher(['service' => [$service, $method]]);
```

But who knows, maybe someone wiil find it useful. So, again: you're the author, you decide :)

**Comment by ksimka** [comment]
So, what about this PR? Can I do anything else?

**Comment by harikt** [comment]
I could not comment on this. Please wait for @pmjones .

**Comment by pmjones** [comment]
After letting it sit in my mind for a while, this still doesn't seem right to me. Sorry @ksimka -- maybe another time.

**Pull request closed without merging**